server/etcdserver/server.go

```diff
@@ -1810,11 +1810,14 @@ func (s *EtcdServer) apply(
 // applyEntryNormal applies an EntryNormal type raftpb request to the EtcdServer
 func (s *EtcdServer) applyEntryNormal(e *raftpb.Entry) {
 	shouldApplyV3 := membership.ApplyV2storeOnly
+	var ar *applyResult
 	applyV3Performed := false
 	defer func() {
 		// The txPostLock callback will not get called in this case,
 		// so we should set the consistent index directly.
-		if s.consistIndex != nil && !applyV3Performed && membership.ApplyBoth == shouldApplyV3 {
+		if s.consistIndex != nil &&
```

Comment:
I'm trying to think whether the whole condition makes sense: 
 - why `s.consistIndex` could be null... seems likely in tests only.
 - We don't set `applyV3Performed=true` if the function does not has side-effects... but if it  has no side-effects... there should nothing wrong with rolling cindex forward as well.. 

What if we set the defer in `if e.Index > index {` branch (1828 line) ?

Reply:
> why s.consistIndex could be null... seems likely in tests only.

yes




> * We don't set `applyV3Performed=true` if the function does not has side-effects... but if it  has no side-effects... there should nothing wrong with rolling cindex forward as well..

I am not sure I got your point. If the  `s.applyV3.Apply` (line 1879) fails, then probably the LockInsideApply isn't called at all. For example, in this case, when the authentication fails.




> What if we set the defer in `if e.Index > index {` branch (1828 line) ?

No, we can't do this. Because we need the new values during the execution of `s.applyV3.Apply` (line 1879) .

Reply:
> > why s.consistIndex could be null... seems likely in tests only.
> 
> yes

So we had batter to not complicate prod code for 'test laziness' and fix the tests...

> 
> > * We don't set `applyV3Performed=true` if the function does not has side-effects... but if it  has no side-effects... there should nothing wrong with rolling cindex forward as well..
> 
> I am not sure I got your point. If the `s.applyV3.Apply` (line 1879) fails, then probably the LockInsideApply isn't called at all. For example, in this case, when the authentication fails.
> 
> > What if we set the defer in `if e.Index > index {` branch (1828 line) ?
> 
> No, we can't do this. Because we need the new values during the execution of `s.applyV3.Apply` (line 1879) .

I'm not saying we should remove
```
s.consistIndex.SetConsistentApplyingIndex(e.Index, e.Term)
```  there. 

I'm proposing to do both:
	if e.Index > index {
		// set the consistent index of current executing entry
		s.consistIndex.SetConsistentApplyingIndex(e.Index, e.Term)
		shouldApplyV3 = membership.ApplyBoth
        defer s.consistIndex.SetConsistentIndex(e.Index, e.Term)
        // or just defer s.consistIndex.Applied() that moves applying to applied.
	}

Reply:
Great proposal on the "`defer s.consistIndex.SetConsistentIndex(e.Index, e.Term)`" !!! It simplifies the implementation!!!

Regarding to the comment "So we had batter to not complicate prod code for 'test laziness' and fix the tests...", I agree with you. I will evaluate the effort, if it affects lots of test, then I tend to fix it in a separate PR. WDYT?

Reply:
Resolved, thanks @ptabor